project_euler/problem_54/sol1.py

```diff
@@ -0,0 +1,366 @@
+        >>> player.compare_with(opponent)
+        'Loss'
+
+        >>> player = PokerHand("2S AH 2H AS AC")  # Full house
+        >>> opponent = PokerHand("2H 3H 5H 6H 7H")  # Flush
+        >>> player.compare_with(opponent)
+        'Win'
+
+        >>> player = PokerHand("2S AH 4H 5S 6C")  # High card
+        >>> opponent = PokerHand("AD 4C 5H 6H 2C")  # High card
+        >>> player.compare_with(opponent)
+        'Tie'
+        """
+        # Breaking the tie works on the following order of precedence:
+        # 1. First pair (default 0)
+        # 2. Second pair (default 0)
+        # 3. Compare all cards in reverse order because they are sorted.
+
+        # First pair and second pair will only be a non-zero value if the card
+        # type is either from the following:
+        # 21: Four of a kind
+        # 20: Full house
+        # 17: Three of a kind
+        # 16: Two pairs
+        # 15: One pair
+        if self._hand_type > other._hand_type:
+            return "Win"
+        elif self._hand_type < other._hand_type:
+            return "Loss"
+        elif self._first_pair == other._first_pair:
+            if self._second_pair == other._second_pair:
+                return self._compare_cards(other)
+            else:
+                return "Win" if self._second_pair > other._second_pair else "Loss"
+        return "Win" if self._first_pair > other._first_pair else "Loss"
+
+    # This function is not part of the problem, I did it just for fun
+    def hand_name(self) -> str:
+        """
+        Return the name of the hand in the following format:
+        'hand name, high card'
+
+        Here are some examples:
+        >>> PokerHand("KS AS TS QS JS").hand_name()
+        'Royal flush'
+
+        >>> PokerHand("2D 6D 3D 4D 5D").hand_name()
+        'Straight flush, Six-high'
+
+        >>> PokerHand("JC 6H JS JD JH").hand_name()
+        'Four of a kind, Jacks'
+
+        >>> PokerHand("3D 2H 3H 2C 2D").hand_name()
+        'Full house, Twos over Threes'
+
+        >>> PokerHand("2H 4D 3C AS 5S").hand_name()  # Low ace
+        'Straight, Five-high'
+
+        Source: https://en.wikipedia.org/wiki/List_of_poker_hands
+        """
+        name = PokerHand._HAND_NAME[self._hand_type - 14]
+        high = PokerHand._CARD_NAME[self._high_card]
+        pair1 = PokerHand._CARD_NAME[self._first_pair]
+        pair2 = PokerHand._CARD_NAME[self._second_pair]
+        if self._hand_type in [22, 19, 18]:
+            return name + f", {high}-high"
+        elif self._hand_type in [21, 17, 15]:
+            return name + f", {pair1}s"
+        elif self._hand_type in [20, 16]:
+            join = "over" if self._hand_type == 20 else "and"
+            return name + f", {pair1}s {join} {pair2}s"
+        elif self._hand_type == 23:
+            return name
+        else:
+            return name + f", {high}"
+
+    def _compare_cards(self, other: "PokerHand") -> str:
+        # Comparing in reverse order as the list is sorted
```

Comment:
Why sort the list backwards?  If it is sorted in the way you want to use it then the logic is easier to follow and we can use enumerate.

Reply:
I didn't sort the list backward. It is sorted in ascending order but in this case, I need the elements in reverse order. When two poker hands are tied you need to check the card values from highest to lowest. I don't think `enumerate` will work here as I am comparing elements from two different lists.

Let me see, I don't think it's feasible as I need it in `_is_five_high_straight` (I use `pop` to remove the last element) and also it's easy to look the hand in ascending order. Still, I'm open to suggestions.

Reply:
> I don't think enumerate will work here as I am comparing elements from two different lists.

enumerate() works fine.  It allows you to replace self.hand.cards[i] with card.

Reply:
Oh right, but I still have to keep `other._cards[i][0]` and `enumerate` only takes a start argument. Also, I won't be able to just put `cards` in place of `self._cards[i][0]` as my internal state of the hand is a list of two tuple `[(value, suit), (value, suit) ...]`. And as I pointed out, I will have to change in a few other places where I'm assuming the list is in ascending order especially in `_is_five_high_straight`.

Oh damn, I got an idea which can simplify a lot of things: I am mostly using the card values except in `_is_flush` to determine whether all suit is the same. I have already separated the values and so what I will do is make a `set` of cards and if the length of `set` is one then it's a flush otherwise it is not. This will remove all the double indexing and I will incorporate `enumerate` into it as well. Thanks for all the suggestions!